onnxscript/type_annotation.py

```diff
@@ -143,6 +166,85 @@ def get_return_types(typeinfo: type | typing.Sequence[type]) -> typing.Sequence[
     """
     if isinstance(typeinfo, typing.Sequence):
         return typeinfo
-    if get_origin(typeinfo) is tuple:
-        return get_args(typeinfo)
+    if typing.get_origin(typeinfo) is tuple:
+        return typing.get_args(typeinfo)
     return (typeinfo,)
+
+
+def pytype_to_input_strings(pytype: TypeAnnotationValue) -> list[str]:
+    """Returns a list of all supported input types in string representation for a given type annotation.
+
+    Args:
+        pytype: A type annotation.
+
+    Returns:
+        A list of all supported input types for the given type annotation.
+        Ensures that the list is sorted in the same order as ALL_TYPE_STRINGS.
+    """
+    if pytype is None:
+        return list(ALL_TENSOR_TYPE_STRINGS)
+    if pytype is onnx_types.TensorType:
+        return list(ALL_TENSOR_TYPE_STRINGS)
+    if isinstance(pytype, type) and issubclass(pytype, onnx_types.TensorType):
+        return [pytype.to_string()]
+    if isinstance(pytype, onnx_types.TensorType):
+        return [pytype.to_string()]
+    if isinstance(pytype, typing.TypeVar):
+        constraints = pytype.__constraints__
+        if constraints:
+            return pytype_to_input_strings(Union.__getitem__(constraints))
+        bound = pytype.__bound__
+        if bound is None:
+            return list(ALL_TENSOR_TYPE_STRINGS)
+        return pytype_to_input_strings(bound)
+    if typing.get_origin(pytype) is Union:
+        options = []
+        subtypes = typing.get_args(pytype)
+        # A None type in a Union is equivalent to an optional type
+        optional = is_optional(pytype)
+        for subtype in subtypes:
+            if subtype is type(None):
+                # Skip None type because we are handling it with is_optional
+                continue
+            if optional:
+                options += [
+                    *pytype_to_input_strings(subtype),
+                    *[f"optional({s})" for s in pytype_to_input_strings(subtype)],
+                ]
+            else:
+                options += pytype_to_input_strings(subtype)
+        # Remove duplicates
+        return sorted(set(options))
+    if typing.get_origin(pytype) in _LIST_CONSTRUCTORS:
+        subtypes = typing.get_args(pytype)
+        return [f"seq({s})" for s in pytype_to_input_strings(subtypes[0])]
+
+    raise ValueError(f"Unsupported type: {pytype}")
+
+
+def get_type_constraint_name(pytype: TypeAnnotationValue) -> Optional[str]:
```

Comment:
When is this function used/useful?

Reply:
It’s used for converting Python function signatures into onnx type constraints. The tests show usage. You can find actual usage in #626 and #674